File: framework/languages/node/e2e/RegistryMessaging/Client/Scenarios/rm-c4-timeout-isolation-scenario.ts
// RM-C4: Timeout 뒤 late reply를 버린다 시나리오를 검증한다.
import type { ProfileRes, RequestFailureRes } from '../../Shared/messages';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

export async function runRmC4(locationConsumerUrl: string, providerAUrl: string, providerBUrl: string): Promise<void> {
  const timeout = await postJson<RequestFailureRes>(locationConsumerUrl, '/profile/slow-request', { value: 'slow' });
  ensure(timeout.failed, 'RM-C4 expected the slow request to time out.');

  const immediate = await postJson<ProfileRes>(locationConsumerUrl, '/profile/request', { value: 'rm-c4-after-timeout' });
  ensure(immediate.value === 'profile:rm-c4-after-timeout', 'RM-C4 follow-up reply mismatch.');

  await new Promise((resolve) => setTimeout(resolve, 1200));
  const later = await postJson<ProfileRes>(locationConsumerUrl, '/profile/request', { value: 'rm-c4-later' });
  ensure(later.value === 'profile:rm-c4-later', 'RM-C4 later reply mismatch.');

  const afterTimeout = await firstEvidence(providerAUrl, providerBUrl, 'rm-c4-after-timeout');
  const laterEvidence = await firstEvidence(providerAUrl, providerBUrl, 'rm-c4-later');
  const lines = [...afterTimeout, ...laterEvidence];
  ensure(lines.some((line) => line.includes('rm-c4-after-timeout')) && lines.some((line) => line.includes('rm-c4-later')), 'RM-C4 follow-up evidence missing.');
  console.log('scenario RM-C4 passed');
}

async function firstEvidence(providerAUrl: string, providerBUrl: string, marker: string): Promise<string[]> {
  return await Promise.race([
    postJson<string[]>(providerAUrl, '/evidence/wait', { contains: marker }),
    postJson<string[]>(providerBUrl, '/evidence/wait', { contains: marker })
  ]);
}

File: framework/languages/dotnet/e2e/LocationMessaging/Client/Scenarios/RmC4TimeoutIsolationScenario.cs
// Verifies RM-C4 Timeout Isolation behavior.
using LocationMessaging.Client.Support;
using LocationMessaging.Shared;
using Zlink.Framework.Contracts.Errors;
using Zlink.HttpClient;

namespace LocationMessaging.Client.Scenarios;

// RM-C4 verifies that a timed-out request does not poison later requests on
// the same location-store auto-connected channel.
internal static class RmC4TimeoutIsolationScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient storeConsumer,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var timeout = (await storeConsumer.Post("/profile/slow-request")
            .Body(new ProfileReq("slow"))
            .Async<ExpectedFailureRes>()).Body;
        ZlinkStreamAssert.Ensure(
            timeout.ErrorKind == nameof(ZLinkFrameworkErrorKind.DeadlineExceeded),
            "RM-C4 expected DeadlineExceeded.");

        var immediate = (await storeConsumer.Post("/profile/request")
            .Body(new ProfileReq("rm-c4-after-timeout"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(immediate.Value == "profile:rm-c4-after-timeout", "RM-C4 follow-up reply mismatch.");

        // The timed-out request may still be in flight (dealer connect or
        // handler completion after the client-side timeout); the completion
        // evidence gets a wider window than the default poll.
        var slowCompletion = await ProviderEvidence.WaitFromEitherAsync(
            providerA, providerB, "value=slow", timeoutMilliseconds: 30000);
        ZlinkStreamAssert.Ensure(
            slowCompletion.Any(line => line.Contains("profile-request", StringComparison.Ordinal)
                                       && line.Contains("value=slow", StringComparison.Ordinal)),
            "RM-C4 timed-out handler did not complete on the server.");

        var later = (await storeConsumer.Post("/profile/request")
            .Body(new ProfileReq("rm-c4-later"))
            .Async<ProfileRes>()).Body;
        ZlinkStreamAssert.Ensure(later.Value == "profile:rm-c4-later", "RM-C4 later reply mismatch.");

        var afterTimeoutEvidence = await ProviderEvidence.WaitFromEitherAsync(
            providerA,
            providerB,
            "rm-c4-after-timeout");
        var laterEvidence = await ProviderEvidence.WaitFromEitherAsync(providerA, providerB, "rm-c4-later");
        var evidence = afterTimeoutEvidence.Concat(laterEvidence).ToArray();
        ZlinkStreamAssert.Ensure(
            evidence.Any(line => line.Contains("rm-c4-after-timeout", StringComparison.Ordinal))
            && evidence.Any(line => line.Contains("rm-c4-later", StringComparison.Ordinal)),
            "RM-C4 follow-up evidence missing.");
    }

}
